docs: make README.md clearer (#142)
Mic92/sops-nix: README.md

![sops-nix logo](https://github.com/Mic92/sops-nix/releases/download/assets/logo.gif "Logo of sops-nix")

Atomic, declarative, and reproducible secret provisioning for NixOS based on [sops](https://github.com/mozilla/sops).

## How it works

Secrets are decrypted from [`sops` files](https://github.com/mozilla/sops#2usage) during

> activation time. The secrets are access-controlled by by full declarative configuration of their users, permissions, and groups.

Reviewer: ```suggestion
activation time. The secrets are stored as one secret per file and access-controlled by full declarative configuration of their users, permissions, and groups.
```

> It can use `age`, GPG, or SSH keys for decryption. 

Reviewer: ```suggestion
It can use `age` or `GPG` for decryption, or derive these keys from rsa or ed25519 SSH keys.
```

> In the future, we will also support cloud key management APIs such as AWS
> KMS, GCP KMS, Azure Key Vault and Hashicorp Vault.

Reviewer: ```suggestion
Sops also supports cloud key management APIs such as AWS
KMS, GCP KMS, Azure Key Vault and Hashicorp Vault. While not
officially supported by sops-nix, these can be controlled using
environment variables that can be passed to sops.
```

Author: Haha, I was going to comment on this :P

## Features

- Compatible with all NixOS deployment frameworks: [NixOps](https://github.com/NixOS/nixops), nixos-rebuild, [krops](https://github.com/krebs/krops/), [morph](https://github.com/DBCDK/morph), [nixus](https://github.com/Infinisil/nixus), etc.

> - Version-control friendly: Since all files are encrypted they can directly committed to version control without worry. Diffs of the secrets are readable, and [can be shown in cleartext](https://github.com/mozilla/sops#showing-diffs-in-cleartext-in-git).

Reviewer: ```suggestion
- Version-control friendly: Since all files are encrypted they can be directly committed to version control without worry. Diffs of the secrets are readable, and [can be shown in cleartext](https://github.com/mozilla/sops#showing-diffs-in-cleartext-in-git).
```

- CI friendly: Since sops files can be added to the Nix store without leaking secrets, a machine definition can be built as a whole from a repository, without needing to rely on external secrets or services.

> - Works well in teams: sops-nix comes with `nix-shell` hooks that allows multiple people to quickly import all keys.

Reviewer: ```suggestion
- Works well in teams: sops-nix comes with `nix-shell` hooks that allows multiple people to quickly import all GPG keys.
```

>   The cryptography used in sops is designed to be scalable: Secrets are only encrypted once with a master key

Author: Is this correct? I can't find it in the sops documentation, and it appears to be wrong by taking a cursory glance at the produced encrypted YAML files: in cases where you specify more than one key, they seem to encrypt the message twice.

Reviewer: The same secret is only encrypted against the master key. The master key itself is encrypted which each recipient so you only have to decrypt it once with your private key

Author: That makes sense.

> $ niv add Mic92/sops-nix

Author: Is this still the current recommendation? [Mic92's personal configuration](https://github.com/Mic92/dotfiles/tree/master/nixos) seems to use flakes (see the `flake.nix` file at the root.)

Reviewer 2: Maybe we can move flakes up and recommend `niv` for people not using flakes?

}
```

</details>

<details>
<summary><b>2. Generate a key for yourself</b></summary>

This key will be used for you to edit secrets.

You can generate yourself a key:
```console
# for age
$ mkdir -p ~/.config/sops/age
$ age-keygen -o ~/.config/sops/age/keys.txt
# to convert an ssh ed25519 key to an age key
$ mkdir -p ~/.config/sops/age
$ nix-shell -p ssh-to-age --run "ssh-to-age -private-key -i ~/.ssh/id_ed25519 > ~/.config/sops/age/keys.txt"
# for GPG >= version 2.1.17
$ gpg --full-generate-key
# for GPG < 2.1.17
$ gpg --default-new-key-algo rsa4096 --gen-key
```

Or you can use the `ssh-to-pgp` tool to get a GPG key from an SSH key: 
```console
$ nix-shell -p gnupg -p ssh-to-pgp --run "ssh-to-pgp -private-key -i $HOME/.ssh/id_rsa | gpg --import --quiet"
2504791468b153b8a3963cc97ba53d1919c5dfd4
# This exports the public key
$ nix-shell -p ssh-to-pgp --run "ssh-to-pgp -i $HOME/.ssh/id_rsa -o $USER.asc"
2504791468b153b8a3963cc97ba53d1919c5dfd4
```

> (Note that `ssh-to-pgp` currently only supports RSA keys.)  

Reviewer: ```suggestion
(Note that `ssh-to-pgp` only supports RSA keys.)  
```

If you get the following,
```console
ssh-to-pgp: failed to parse private ssh key: ssh: this private key is passphrase protected
```
then your SSH key is encrypted with your password and you will need to create an unencrypted copy temporarily.
```console
$ cp $HOME/.ssh/id_rsa /tmp/id_rsa
$ ssh-keygen -p -N "" -f /tmp/id_rsa
$ nix-shell -p gnupg -p ssh-to-pgp --run "ssh-to-pgp -private-key -i /tmp/id_rsa | gpg --import --quiet"
$ rm /tmp/id_rsa
```

> You can also use your existing SSH key as an age key.

Reviewer: ```suggestion
You can also use your existing ed25519 SSH key as an age key.
```

/tmp/tmp.JA07D1aVRD/pubring.kbx
-------------------------------
sec   rsa2048 1970-01-01 [SCE]
      9F89C5F69A10281A835014B09C3DC61F752087EF
uid           [ unknown] root <root@localhost>
```

The fingerprint here is `9F89C5F69A10281A835014B09C3DC61F752087EF`.
</details>

Your age public key or GPG fingerprint can written to your [`.sops.yaml`](https://github.com/mozilla/sops#using-sops-YAML-conf-to-select-kms-pgp-for-new-files) in the root of your configuration directory or repository:
```yaml
# This example uses YAML anchors which allows reuse of multiple keys 
# without having to repeat yourself.
# Also see https://github.com/Mic92/dotfiles/blob/master/nixos/.sops.yaml
# for a more complex example.
keys:
  - &admin_alice 2504791468b153b8a3963cc97ba53d1919c5dfd4
  - &admin_bob age12zlz6lvcdk6eqaewfylg35w0syh58sm7gh53q5vvn7hd7c6nngyseftjxl
creation_rules:
  - path_regex: secrets/[^/]+\.yaml$
    key_groups:
    - pgp:
      - *admin_alice
    - age:
      - *admin_bob
```

</details>

<details>
<summary><b>3. Get a public key for your target machine</b></summary>

The easiest way to add new machines is using SSH host keys (this requires OpenSSH to be enabled).  

> If you are using age, the `ssh-to-age` tool can be used to convert any SSH public key to the age format:

Reviewer: ```suggestion
If you are using age, the `ssh-to-age` tool can be used to convert any ed25519 SSH public key to the age format:
```

```console
$ nix-shell -p ssh-to-age --run 'ssh-keyscan my-server.com | ssh-to-age'
age1rgffpespcyjn0d8jglk7km9kfrfhdyev6camd3rck6pn8y47ze4sug23v3
$ nix-shell -p ssh-to-age --run 'cat /etc/ssh/ssh_host_ed25519_key.pub | ssh-to-age'
age1rgffpespcyjn0d8jglk7km9kfrfhdyev6camd3rck6pn8y47ze4sug23v3
```

> For GPG, since sops does not natively support SSH keys yet, nix-sops supports a conversion tool (`ssh-to-pgp`)

Reviewer: ```suggestion
For GPG, since sops does not natively support SSH keys yet, sops-nix supports a conversion tool (`ssh-to-pgp`)
```

## Different file formats

At the moment we support the following file formats: YAML, JSON, and binary.

NOTE: At the moment we do not support nested data structures that
sops support. This might change in the future.

> We support the following YAML:

Author: I have no idea what this means. Isn't nested YAML already supported by subdirectories? Or am I missing something?

Reviewer: I forgot one passage about nested keys when I implemented the subdirectory feature and it still mentions that that's not supported. Maybe you mean this?

Author: Should I remove this part, then?

Reviewer 2: Yes. This should be dropped.

  imports = [ <sops-nix/modules/sops> ];
  # The default sops file used for all secrets can be controlled using `sops.defaultSopsFile`
  sops.defaultSopsFile = ./secrets.yaml;
  # If you use something different from YAML, you can also specify it here:

>   #sops.defaultSopsFormat = "YAML";

Reviewer 2: ```suggestion
  #sops.defaultSopsFormat = "yaml";
```

```nix
{
  sops.defaultSopsFile = ./secrets.yaml;
  # YAML is the default 

>   #sops.defaultSopsFormat = "YAML";

Reviewer 2: ```suggestion
  #sops.defaultSopsFormat = "yaml";
```

  sops.secrets.github_token = {

>     format = "YAML";

Reviewer 2: ```suggestion
    format = "yaml";
```

## Migrate from pass/krops

If you have used [pass](https://www.passwordstore.org) before (e.g. in
[krops](https://github.com/krebs/krops)) than you can use the following one-liner
to convert all your secrets to a YAML structure:

```console

> $ for i in *.gpg; do echo "$(basename $i .gpg): |\n$(pass $(dirname $i)/$(basename $i .gpg)| sed 's/^/  /')"; done

Author: This one-liner doesn't work; it doesn't handle nested directories, which show up commonly if one is using `pass`. To be honest, I'm not entirely sure it's easy to come up with a one-liner for this because of how convoluted the YAML format is (how do we handle special characters e.g. `\0` or `\r`?) The easiest approach I can think of is somehow encoding to JSON, then using `yq -P` to translate to idiomatic YAML.

Reviewer 2: This was only used for krops at a time where there was no support for nesting anyway.

Reviewer 2: We could also drop this otherwise. I don't think there are too many krops user, that not migrated yet.

Reviewer 2: Let's keep this for another PR.
